nodebuilder/node/uptime.go

```diff
@@ -0,0 +1,83 @@
+// This file defines UptimeMetrics relative to the nodebuilder package.
+package node
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
+	"go.opentelemetry.io/otel/metric/instrument/syncfloat64"
+)
+
+// UptimeMetrics is a struct that records
+// 1. node start time: the timestamp when the node was started
+// 2. node up time: total time the node has been running
+//
+// the node start time is recorded imperatively when RecordNodeStartTime is called
+// whereas the node up time is recorded periodically upon callback recalling (re-mettering from optl)
+type UptimeMetrics struct {
+	// nodeStartTS is the timestamp when the node was started.
+	nodeStartTS syncfloat64.UpDownCounter
+
+	// totalNodeUptime is the total time the node has been running.
+	totalNodeUptime asyncfloat64.Gauge
+
+	// lastNodeUptimeTS is the last timestamp when the node uptime was recorded.
+	lastNodeUptimeTS float64
+}
+
+var meter = global.MeterProvider().Meter("node")
+
+// NewUptimeMetrics creates a new UptimeMetrics
+// and registers a callback to re-meter the totalNodeUptime metric.
+func NewUptimeMetrics() (*UptimeMetrics, error) {
+	nodeStartTS, err := meter.
```

Comment:
Does this store the node start value in some sort of db and retrieve it if available? because if it doesn't, once the node is restarted, we get a new start time for the node which yields in a wrong uptime calculation.

Reply:
Good point, now implemented a mechanism to store it in the datastore in the [following commit](https://github.com/celestiaorg/celestia-node/pull/1638/commits/f01f9281292bc8926691f98e330177c33b8a4bda)